src/datastore/postgres/PgDatastore.ts

```diff
@@ -403,11 +403,9 @@ export class PgDatastore implements Datastore {
         throw Error("Couldn't fetch schema version");
     }
 
-    private static BuildUpsertStatement(table: string, conflictKey: string, keyValues: {[key: string]: string}) {
-        const keys = Object.keys(keyValues).join(", ");
-        const keysValues = `\${${Object.keys(keyValues).join("}, ${")}}`;
-        // tslint:disable-next-line: prefer-template
-        const keysSets = Object.keys(keyValues).slice(1).map((k) => `${k} = \${${k}}`).join(", ");
-        return `INSERT INTO ${table} (${keys}) VALUES (${keysValues}) ON CONFLICT (${conflictKey}) DO UPDATE SET ${keysSets}`;
+    private static BuildUpsertStatement(table: string, conflictKeys: string[], values: Array<{[key: string]: any}>) {
+        const cs = new pgp.helpers.ColumnSet(values[0], {table});
```

Comment:
Is it a problem if `values[0]` contains less columns than `values[1]`?

Reply:
Yes, it is. I generally do not recommend use of an object to extract column names. That's what I meant by:

> This will solve about 90% of problems, creating room for further improvements.

A better approach is to pre-declare each `ColumnSet`, specifying column names statically.

Reply:
> Looks redundant to me. 

Not quite. You'd need to add static ColumnSet, and make use of it, then it will be redundant and can be refactored. In the current version however the code relies on column names within the data object.

Reply:
@jaller94 You will do well, simply following the logic shown in [Multi-row insert with pg-promise](https://stackoverflow.com/questions/37300997/multi-row-insert-with-pg-promise) ;)